buildSrc/src/main/java/org/elasticsearch/gradle/internal/rest/compat/RestCompatTestTransformTask.java

```diff
@@ -0,0 +1,151 @@
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+public class RestCompatTestTransformTask extends DefaultTask {
+
+    private static final YAMLFactory yaml = new YAMLFactory();
+    private static final ObjectMapper mapper = new ObjectMapper(yaml);
+
+    private static final Map<String, String> headers = Map.of(
+        "Content-Type",
+        "application/vnd.elasticsearch+json;compatible-with=7",
+        "Accept",
+        "application/vnd.elasticsearch+json;compatible-with=7"
+    );
+
+    private String sourceSetName;
+    private static final String REST_TEST_PREFIX = "rest-api-spec/test";
+
+    private final PatternFilterable testPatternSet;
+    private final List<RestTestTransform<?>> transformations;
+
+    private String inputResourceParent = "";
+
+    public RestCompatTestTransformTask() {
+        testPatternSet = getPatternSetFactory().create();
+        testPatternSet.include("/*" + "*/*.yml"); // concat these strings to keep build from thinking this is invalid javadoc
+        transformations = Collections.singletonList(new InjectHeaders(headers));
+    }
+
+    @Inject
+    protected Factory<PatternSet> getPatternSetFactory() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Input
+    public String getSourceSetName() {
+        return sourceSetName;
+    }
+
+    @OutputDirectory
+    public File getOutputDir() {
+        return new File(
+            getSourceSet().orElseThrow(() -> new IllegalArgumentException("could not find source set [" + sourceSetName + "]"))
+                .getOutput()
+                .getResourcesDir(),
+            REST_TEST_PREFIX
+        );
+    }
+
+    @SkipWhenEmpty
+    @InputFiles
+    public FileTree getTestFiles() {
+        return getProject().files(
+            new File(
+                new File(
+                    getSourceSet().orElseThrow(() -> new IllegalArgumentException("could not find source set [" + sourceSetName + "]"))
+                        .getOutput()
+                        .getResourcesDir(),
+                    inputResourceParent
+                ),
+                REST_TEST_PREFIX
+            )
+        ).getAsFileTree().matching(testPatternSet);
+    }
+
+    @TaskAction
+    public void transform() throws IOException {
+        RestTestTransformer transformer = new RestTestTransformer();
+        for (File file : getTestFiles().getFiles()) {
+            YAMLParser yamlParser = yaml.createParser(file);
+            List<ObjectNode> tests = mapper.readValues(yamlParser, ObjectNode.class).readAll();
+            List<ObjectNode> transformRestTests = transformer.transformRestTests(new LinkedList<>(tests), transformations);
+            // convert to url to ensure forward slashes
+            String[] testFileParts = file.toURI().toURL().getPath().split(REST_TEST_PREFIX);
+            assert testFileParts.length == 2; // before and after the REST_TEST_PREFIX
```

Comment:
Let's not use assertions here. I'd prefer we throw proper exceptions.

Reply:
done.  (and the other assert shouldn't have been, so i removed it)